bsm powermeter

chrkli22 opened the pull request:
- implements powermeter interface, including "get_signed_meter_value", which returns a signed OCMF snapshot string.
- publishes variable "powermeter" of type Powermeter.
- type Powermeter: added optional VAR property.
- cmd for requesting standard sunspec ac meter model.
- uses SerialCommHub or direct modbus.

Module currently runs without problems on its testinstallation.

Branch will be squashed when this pull request is accepted.

a-w50 commented on modules/PowermeterBSM/lib/sunspec_base.hpp:
```diff
@@ -0,0 +1,309 @@
+};
+// clang-format on
+
+static std::string pointTypeToString(PointType pointType) {
```
casing

a-w50 commented on modules/PowermeterBSM/lib/sunspec_base.hpp:
```diff
@@ -0,0 +1,309 @@
+    return it == pointTypeToStringMap.end() ? "unknown type" : it->second;
+}
+
+constexpr std::uint16_t point_length_in_bytes(PointType pointType) {
```
casing

a-w50 commented on modules/PowermeterBSM/lib/sunspec_base.hpp:
```diff
@@ -0,0 +1,309 @@
+
+struct Point {
+    const char* id;
+    PointType pointType;
```
casing

chrkli22 replied:
fixed

a-w50 commented on modules/PowermeterBSM/lib/sunspec_base.hpp:
```diff
@@ -0,0 +1,309 @@
+namespace SunspecDecoder {
+
+inline std::uint16_t uint16_at(const Types::DataVector& data, Types::DataVector::size_type offset) {
+    return be16toh((*reinterpret_cast<const std::uint16_t*>(std::addressof(data.data()[offset]))));
```
what about (data.data() + offset) instead of std::addressof(data.data()[offset])?

a-w50 reviewed (changes requested):
- source files missing copyright
- include guards quite short (ifdef collision)
- multiline comments deprecated
- look at codacy issues
- namespaces should be lowercase
- base_types / known_models / sunspec_* / driver_base belong into libmodbus/libsunspec
- only transport and BSM... will stay, for these you wouldn't need all this cmake extra stuff if you add them directly as sources to the module

chrkli22 commented:
> * source files missing copyright
fixed.

> * include guards quite short (ifdef collision)
never seen this happen, changed that anyway.

> * multiline comments deprecated
Hu? Who deprecated, and why?

> * look at codacy issues
> * namespaces should be lowercase
fixed.

> * base_types / known_models / sunspec_* / driver_base belong into libmodbus/libsunspec
No, libsunspec is not used here. 

> * only transport and BSM... will stay, for these you wouldn't need all this cmake extra stuff if you add them directly as sources to the module

shortened that cmake stuff a bit.

chrkli22 commented on modules/PowermeterBSM/manifest.json:
```diff
@@ -0,0 +1,73 @@
+			"minimum": 1
+		},
+		"watchdog_wakeup_interval" : {
+			"description": "wakup interval of watchdog in seconds (default 60 seconds).",
```
This would sound like "wake up the module", but this is paramter how the module works.

chrkli22 commented on modules/PowermeterBSM/manifest.json:
```diff
@@ -0,0 +1,73 @@
+	"requires": {
+		"serial_com_0_connection": {
+			"interface": "serial_communication_hub",
+			"min_connections": 1,
```
Yes, valid point. I did not find a way to make this optional.

chrkli22 commented on modules/PowermeterBSM/CMakeLists.txt:
```diff
@@ -0,0 +1,64 @@
+target_include_directories( sunspec_framework_object_lib PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} )
+target_include_directories( sunspec_framework_object_lib PUBLIC ${CMAKE_BINARY_DIR}/generated/include )
+target_include_directories( sunspec_framework_object_lib PUBLIC ${CMAKE_BINARY_DIR}/generated/modules/${MODULE_NAME} )
+target_link_libraries( sunspec_framework_object_lib PUBLIC everest::framework )
```
cleaned that up.

chrkli22 commented on modules/PowermeterBSM/CMakeLists.txt:
```diff
@@ -0,0 +1,64 @@
+target_include_directories( sunspec_framework_object_lib PUBLIC ${CMAKE_BINARY_DIR}/generated/modules/${MODULE_NAME} )
+target_link_libraries( test_rtu_device PUBLIC
+    sunspec_framework_object_lib
+    GTest::gtest_main
```
building tests now is optional.

chrkli22 commented on modules/PowermeterBSM/lib/transport.hpp:
```diff
@@ -0,0 +1,119 @@
+    ModbusTransport(std::string serial_device_name, Types::ModbusUnitId unit_id,
+                    Types::ModbusRegisterAddress base_address);
+
+    inline everest::connection::RTUConnection& rtu_connection() {
```
I changed it. I learned that the meaning of "inline" has changed a bit.

chrkli22 commented on modules/PowermeterBSM/lib/transport.cpp:
```diff
@@ -0,0 +1,189 @@
+
+    std::uint16_t response_status = 42;
+
+    std::size_t counter = 10;
```
refactored to const values, but not in extra headers, since these values only make sense there.

chrkli22 commented on modules/PowermeterBSM/lib/base_types.hpp:
```diff
@@ -0,0 +1,58 @@
+} // namespace Types
+
+Types::SunspecDataModelAddress operator"" _sma(unsigned long long int v);
+Types::ModbusRegisterAddress operator"" _mra(unsigned long long int v);
```
https://en.cppreference.com/w/cpp/language/user_literal

chrkli22 commented on modules/PowermeterBSM/lib/known_model.hpp:
```diff
@@ -0,0 +1,22 @@
+#include "BSMSnapshotModel.hpp"
+#include "sunspec_models.hpp" // sunspec models
+
+namespace KnownModel {
```
added a comment that hopefully explains this.

SebaLukas commented on CMakeLists.txt:
```diff
@@ -22,10 +22,10 @@ if (NOT everest-cmake_FOUND)
     set(everest-cmake_DIR "${everest-cmake_SOURCE_DIR}")
     set(everest-cmake_FIND_COMPONENTS "bundling")
-    include("${everest-cmake_SOURCE_DIR}/everest-cmake-config.cmake")
+    include("${everest-cmake_SOURCE_DIR}/everest-cmake-coes/Powermetefig.cmake")
```
Revert back to everest-cmake-config.cmake

SebaLukas commented on cmake/assets/logging.ini:
```diff
@@ -3,7 +3,10 @@
-Filter="%Severity% >= INFO"
+# Filter="%Process% contains bsm_power_meter"
+Filter="%Severity% > INFO  or %Process% contains bsm_power_meter"
+# Filter="%Severity% >= DEBG"
```
Please delete comments and revert Filter back without contains bsm_power_meter

SebaLukas commented on config/config-sil.json:
```diff
@@ -1,5 +1,13 @@
+        "config_module": {
+            "power_unit_id" : 42,
+            "update_interval" : 5,
+            "serial_device" : "/dev/ttyUSB0"
```
A SIL with a HW dependency is not a SIL. Perhaps a little BSM_Simulator is better here like the slac or imd simulator

SebaLukas commented on interfaces/sunspec_ac_meter.json:
```diff
@@ -0,0 +1,21 @@
+{
+    "description": "get sunspec ac meter measurement",
```
Get

SebaLukas commented on interfaces/sunspec_ac_meter.json:
```diff
@@ -0,0 +1,21 @@
+                }
+            },
+            "result": {
+                "description": "sunspec ac meter model",
```
Sunspec

SebaLukas commented:
- SIL is not running because of missing HW -> SIL should work without HW 
- Codacy issues
- The descriptions are not always capitalized

SebaLukas commented:
There are still a few json files. Do they have to stay in?
